community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -413,12 +466,35 @@ public void close() throws IOException
     @Override
     public IndexWriter<KEY,VALUE> writer( IndexWriter.Options options ) throws IOException
     {
-        SingleIndexWriter result = this.writer.getAndSet( null );
-        if ( result == null )
+        if ( !writerTaken.compareAndSet( false, true ) )
         {
             throw new IllegalStateException( "Only supports one concurrent writer" );
         }
-        return result.take( rootId, options );
+
+        writerCheckpointMutex.lock();
+        boolean success = false;
+        try
+        {
+            writer.take( rootId, options );
+            success = true;
+            return writer;
+        }
+        finally
+        {
+            if ( !success )
+            {
+                releaseWriter();
+            }
```

Comment:
I think it'll call `releaseWriter` if it doesn't succeed and `releaseWriter` will unlock the mutex, won't it?